File: neoforge/src/main/java/net/shiroha233/roadweaver/network/neoforge/MapNetworkForge.java
package net.shiroha233.roadweaver.network.neoforge;

import net.minecraft.client.Minecraft;
import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.network.protocol.common.custom.CustomPacketPayload;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.network.chat.Component;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.world.level.levelgen.Heightmap;
import net.minecraft.core.BlockPos;
import net.neoforged.bus.api.IEventBus;
import net.neoforged.neoforge.network.PacketDistributor;
import net.neoforged.neoforge.network.event.RegisterPayloadHandlersEvent;
import net.neoforged.neoforge.network.registration.PayloadRegistrar;
import net.shiroha233.roadweaver.RoadWeaver;
import net.shiroha233.roadweaver.client.map.RoadMapScreen;
import net.shiroha233.roadweaver.client.map.data.MapDataCollector;
import net.shiroha233.roadweaver.client.map.data.MapSnapshot;
import net.shiroha233.roadweaver.helpers.LevelCompat;
import net.shiroha233.roadweaver.network.MapSnapshotCodec;
import net.shiroha233.roadweaver.helpers.Records;
import net.shiroha233.roadweaver.persistence.WorldDataProvider;
import net.shiroha233.roadweaver.util.ComputeService;

import java.util.ArrayList;
import java.util.concurrent.CompletableFuture;

public class MapNetworkForge {
    public static final ResourceLocation REQ_RECT_ID = ResourceLocation.fromNamespaceAndPath(RoadWeaver.MOD_ID, "map_request_rect");
    public static final ResourceLocation SNAP_ID = ResourceLocation.fromNamespaceAndPath(RoadWeaver.MOD_ID, "map_snapshot");
    public static final ResourceLocation TP_REQ_ID = ResourceLocation.fromNamespaceAndPath(RoadWeaver.MOD_ID, "map_teleport");
    public static final ResourceLocation TP_ACK_ID = ResourceLocation.fromNamespaceAndPath(RoadWeaver.MOD_ID, "map_teleport_ack");
    public static final ResourceLocation MAN_REQ_ID = ResourceLocation.fromNamespaceAndPath(RoadWeaver.MOD_ID, "map_manual_connect");

    public static void register(IEventBus modBus) {
        modBus.addListener(MapNetworkForge::registerPayloads);
    }

    private static void registerPayloads(RegisterPayloadHandlersEvent event) {
        final PayloadRegistrar registrar = event.registrar("1");
        
        registrar.playToClient(MapSnapshotS2C.TYPE, MapSnapshotS2C.CODEC, MapSnapshotS2C::handle);
        registrar.playToClient(TeleportAckS2C.TYPE, TeleportAckS2C.CODEC, TeleportAckS2C::handle);
        
        registrar.playToServer(RequestMapSnapshotC2S.TYPE, RequestMapSnapshotC2S.CODEC, RequestMapSnapshotC2S::handle);
        registrar.playToServer(TeleportC2S.TYPE, TeleportC2S.CODEC, TeleportC2S::handle);
        registrar.playToServer(ManualConnectC2S.TYPE, ManualConnectC2S.CODEC, ManualConnectC2S::handle);
    }

    public static void requestSnapshot(int minX, int minZ, int maxX, int maxZ) {
        sendToServer(new RequestMapSnapshotC2S(minX, minZ, maxX, maxZ));
    }

    public static void requestTeleport(int x, int y, int z) {
        sendToServer(new TeleportC2S(x, y, z));
    }

    public static void requestManualConnect(int ax, int az, int bx, int bz) {
        sendToServer(new ManualConnectC2S(ax, az, bx, bz));
    }

    private static void sendToServer(CustomPacketPayload payload) {
        try {
            var m = PacketDistributor.class.getMethod("sendToServer", CustomPacketPayload.class);
            m.invoke(null, payload);
            return;
        } catch (Throwable ignored) {
        }

        try {
            // Newer PacketDistributor APIs may expose a SERVER target with a noArg() distributor.
            var serverField = PacketDistributor.class.getField("SERVER");
            Object server = serverField.get(null);
            var noArg = server.getClass().getMethod("noArg");
            Object distributor = noArg.invoke(server);
            var send = distributor.getClass().getMethod("send", CustomPacketPayload.class);
            send.invoke(distributor, payload);
        } catch (Throwable t) {
            RoadWeaver.getLogger().warn("Failed to send payload to server: {}", payload.type().id(), t);
        }
    }

    // Payloads

    public record RequestMapSnapshotC2S(int minX, int minZ, int maxX, int maxZ) implements CustomPacketPayload {
        public static final CustomPacketPayload.Type<RequestMapSnapshotC2S> TYPE = new CustomPacketPayload.Type<>(REQ_RECT_ID);
        public static final StreamCodec<FriendlyByteBuf, RequestMapSnapshotC2S> CODEC = StreamCodec.of(
            (buf, v) -> { buf.writeVarInt(v.minX); buf.writeVarInt(v.minZ); buf.writeVarInt(v.maxX); buf.writeVarInt(v.maxZ); },
            buf -> new RequestMapSnapshotC2S(buf.readVarInt(), buf.readVarInt(), buf.readVarInt(), buf.readVarInt())
        );
        @Override public Type<? extends CustomPacketPayload> type() { return TYPE; }
        
        public static void handle(RequestMapSnapshotC2S payload, net.neoforged.neoforge.network.handling.IPayloadContext context) {
            ServerPlayer sp = (ServerPlayer) context.player();
            int cx = (int) Math.round(sp.getX());
            int cz = (int) Math.round(sp.getZ());
            int radiusChunks;
            try {
                net.shiroha233.roadweaver.config.ModConfig cfg = net.shiroha233.roadweaver.config.ConfigService.get();
                radiusChunks = (cfg.dynamicPlanEnabled() ? cfg.dynamicPlanRadiusChunks() : cfg.initialPlanRadiusChunks());
            } catch (Throwable t) {
                radiusChunks = 256;
            }
            int radiusBlocks = Math.max(1, radiusChunks) * 16;
            CompletableFuture
                .supplyAsync(() -> MapDataCollector.build(LevelCompat.getServerLevel(sp), payload.minX, payload.minZ, payload.maxX, payload.maxZ, cx, cz, radiusBlocks), ComputeService.executor())
                .thenAccept(snapshot -> context.enqueueWork(() -> PacketDistributor.sendToPlayer(sp, new MapSnapshotS2C(snapshot))));
        }
    }

    public record MapSnapshotS2C(MapSnapshot snapshot) implements CustomPacketPayload {
        public static final CustomPacketPayload.Type<MapSnapshotS2C> TYPE = new CustomPacketPayload.Type<>(SNAP_ID);
        public static final StreamCodec<FriendlyByteBuf, MapSnapshotS2C> CODEC = StreamCodec.of(
            (buf, v) -> MapSnapshotCodec.write(buf, v.snapshot),
            buf -> new MapSnapshotS2C(MapSnapshotCodec.read(buf))
        );
        @Override public Type<? extends CustomPacketPayload> type() { return TYPE; }
        
        public static void handle(MapSnapshotS2C payload, net.neoforged.neoforge.network.handling.IPayloadContext context) {
            context.enqueueWork(() -> {
                Minecraft mc = Minecraft.getInstance();
                if (mc.screen instanceof RoadMapScreen screen) {
                    screen.setSnapshot(payload.snapshot);
                }
            });
        }
    }

    public record TeleportC2S(int x, int y, int z) implements CustomPacketPayload {
        public static final CustomPacketPayload.Type<TeleportC2S> TYPE = new CustomPacketPayload.Type<>(TP_REQ_ID);
        public static final StreamCodec<FriendlyByteBuf, TeleportC2S> CODEC = StreamCodec.of(
            (buf, v) -> { buf.writeVarInt(v.x); buf.writeVarInt(v.y); buf.writeVarInt(v.z); },
            buf -> new TeleportC2S(buf.readVarInt(), buf.readVarInt(), buf.readVarInt())
        );
        @Override public Type<? extends CustomPacketPayload> type() { return TYPE; }
        
        public static void handle(TeleportC2S payload, net.neoforged.neoforge.network.handling.IPayloadContext context) {
            context.enqueueWork(() -> {
                ServerPlayer sp = (ServerPlayer) context.player();
                boolean allowed = sp.isCreative() || sp.hasPermissions(2);
                if (!allowed) {
                    PacketDistributor.sendToPlayer(sp, new TeleportAckS2C(false, 0, 0, 0));
                    return;
                }
                ServerLevel level = LevelCompat.getServerLevel(sp);
                level.getChunk(payload.x >> 4, payload.z >> 4);
                int ty = level.getHeight(Heightmap.Types.MOTION_BLOCKING_NO_LEAVES, payload.x, payload.z);
                if (ty <= level.getMinY()) ty = level.getSeaLevel() + 1; else ty += 1;
                boolean okTp = LevelCompat.teleport(sp, level, payload.x + 0.5, ty, payload.z + 0.5, sp.getYRot(), sp.getXRot());
                PacketDistributor.sendToPlayer(sp, new TeleportAckS2C(okTp, payload.x, ty, payload.z));
            });
        }
    }

    public record TeleportAckS2C(boolean ok, int x, int y, int z) implements CustomPacketPayload {
        public static final CustomPacketPayload.Type<TeleportAckS2C> TYPE = new CustomPacketPayload.Type<>(TP_ACK_ID);
        public static final StreamCodec<FriendlyByteBuf, TeleportAckS2C> CODEC = StreamCodec.of(
            (buf, v) -> { buf.writeBoolean(v.ok); buf.writeVarInt(v.x); buf.writeVarInt(v.y); buf.writeVarInt(v.z); },
            buf -> new TeleportAckS2C(buf.readBoolean(), buf.readVarInt(), buf.readVarInt(), buf.readVarInt())
        );
        @Override public Type<? extends CustomPacketPayload> type() { return TYPE; }
        
        public static void handle(TeleportAckS2C payload, net.neoforged.neoforge.network.handling.IPayloadContext context) {
            context.enqueueWork(() -> {
                Minecraft mc = Minecraft.getInstance();
                if (mc.player == null) return;
                if (payload.ok) mc.player.displayClientMessage(Component.translatable("gui.roadweaver.map.teleport.success_pos", payload.x, payload.y, payload.z), true);
                else mc.player.displayClientMessage(Component.translatable("gui.roadweaver.map.teleport.denied"), true);
            });
        }
    }

    public record ManualConnectC2S(int ax, int az, int bx, int bz) implements CustomPacketPayload {
        public static final CustomPacketPayload.Type<ManualConnectC2S> TYPE = new CustomPacketPayload.Type<>(MAN_REQ_ID);
        public static final StreamCodec<FriendlyByteBuf, ManualConnectC2S> CODEC = StreamCodec.of(
            (buf, v) -> { buf.writeVarInt(v.ax); buf.writeVarInt(v.az); buf.writeVarInt(v.bx); buf.writeVarInt(v.bz); },
            buf -> new ManualConnectC2S(buf.readVarInt(), buf.readVarInt(), buf.readVarInt(), buf.readVarInt())
        );
        @Override public Type<? extends CustomPacketPayload> type() { return TYPE; }
        
        public static void handle(ManualConnectC2S payload, net.neoforged.neoforge.network.handling.IPayloadContext context) {
            context.enqueueWork(() -> {
                ServerPlayer sp = (ServerPlayer) context.player();
                ServerLevel level = LevelCompat.getServerLevel(sp);
                WorldDataProvider provider = WorldDataProvider.getInstance();
                java.util.List<Records.StructureConnection> origin = provider.getStructureConnections(level);
                java.util.List<Records.StructureConnection> list = origin != null ? new ArrayList<>(origin) : new ArrayList<>();
                BlockPos a = new BlockPos(payload.ax, 0, payload.az);
                BlockPos b = new BlockPos(payload.bx, 0, payload.bz);
                boolean exists = false;
                for (Records.StructureConnection c : list) {
                    BlockPos f = c.from();
                    BlockPos t = c.to();
                    if ((f.equals(a) && t.equals(b)) || (f.equals(b) && t.equals(a))) { exists = true; break; }
                }
                if (!exists) {
                    list.add(new Records.StructureConnection(a, b, Records.ConnectionStatus.PLANNED));
                    provider.setStructureConnections(level, list);
                }
            });
        }
    }
}

File: neoforge/src/main/kotlin/net/shiroha233/roadweaver/network/neoforge/MapNetworkForge.kt
package net.shiroha233.roadweaver.network.neoforge

import net.minecraft.client.Minecraft
import net.minecraft.core.BlockPos
import net.minecraft.network.FriendlyByteBuf
import net.minecraft.network.chat.Component
import net.minecraft.network.codec.StreamCodec
import net.minecraft.network.protocol.common.custom.CustomPacketPayload
import net.minecraft.resources.ResourceLocation
import net.minecraft.server.level.ServerPlayer
import net.minecraft.world.level.levelgen.Heightmap
import net.neoforged.bus.api.IEventBus
import net.neoforged.neoforge.network.PacketDistributor
import net.neoforged.neoforge.network.event.RegisterPayloadHandlersEvent
import net.neoforged.neoforge.network.handling.IPayloadContext
import net.neoforged.neoforge.network.registration.PayloadRegistrar
import net.shiroha233.roadweaver.RoadWeaver
import net.shiroha233.roadweaver.client.map.RoadMapScreen
import net.shiroha233.roadweaver.client.map.data.MapDataCollector
import net.shiroha233.roadweaver.client.map.data.MapSnapshot
import net.shiroha233.roadweaver.helpers.Records
import net.shiroha233.roadweaver.network.MapSnapshotCodec
import net.shiroha233.roadweaver.persistence.WorldDataProvider
import net.shiroha233.roadweaver.util.ComputeService
import java.util.concurrent.CompletableFuture

/**
 * NeoForge 地图网络通道
 *
 * 注意：这里严格保持与 Java 版协议一致（ID、Codec、处理逻辑），避免客户端/服务端不兼容。
 */
@Suppress("MemberVisibilityCanBePrivate")
object MapNetworkForge {
    @JvmField
    val REQ_RECT_ID: ResourceLocation = ResourceLocation.fromNamespaceAndPath(RoadWeaver.MOD_ID, "map_request_rect")

    @JvmField
    val SNAP_ID: ResourceLocation = ResourceLocation.fromNamespaceAndPath(RoadWeaver.MOD_ID, "map_snapshot")

    @JvmField
    val TP_REQ_ID: ResourceLocation = ResourceLocation.fromNamespaceAndPath(RoadWeaver.MOD_ID, "map_teleport")

    @JvmField
    val TP_ACK_ID: ResourceLocation = ResourceLocation.fromNamespaceAndPath(RoadWeaver.MOD_ID, "map_teleport_ack")

    @JvmField
    val MAN_REQ_ID: ResourceLocation = ResourceLocation.fromNamespaceAndPath(RoadWeaver.MOD_ID, "map_manual_connect")

    @JvmStatic
    fun register(modBus: IEventBus) {
        modBus.addListener(::registerPayloads)
    }

    private fun registerPayloads(event: RegisterPayloadHandlersEvent) {
        val registrar: PayloadRegistrar = event.registrar("1")

        registrar.playToClient(MapSnapshotS2C.TYPE, MapSnapshotS2C.CODEC, MapSnapshotS2C::handle)
        registrar.playToClient(TeleportAckS2C.TYPE, TeleportAckS2C.CODEC, TeleportAckS2C::handle)

        registrar.playToServer(RequestMapSnapshotC2S.TYPE, RequestMapSnapshotC2S.CODEC, RequestMapSnapshotC2S::handle)
        registrar.playToServer(TeleportC2S.TYPE, TeleportC2S.CODEC, TeleportC2S::handle)
        registrar.playToServer(ManualConnectC2S.TYPE, ManualConnectC2S.CODEC, ManualConnectC2S::handle)
    }

    @JvmStatic
    fun requestSnapshot(minX: Int, minZ: Int, maxX: Int, maxZ: Int) {
        PacketDistributor.sendToServer(RequestMapSnapshotC2S(minX, minZ, maxX, maxZ))
    }

    @JvmStatic
    fun requestTeleport(x: Int, y: Int, z: Int) {
        PacketDistributor.sendToServer(TeleportC2S(x, y, z))
    }

    @JvmStatic
    fun requestManualConnect(ax: Int, az: Int, bx: Int, bz: Int) {
        PacketDistributor.sendToServer(ManualConnectC2S(ax, az, bx, bz))
    }

    // Payloads

    data class RequestMapSnapshotC2S(val minX: Int, val minZ: Int, val maxX: Int, val maxZ: Int) : CustomPacketPayload {
        override fun type(): CustomPacketPayload.Type<out CustomPacketPayload> = TYPE

        companion object {
            @JvmField
            val TYPE: CustomPacketPayload.Type<RequestMapSnapshotC2S> = CustomPacketPayload.Type(REQ_RECT_ID)

            @JvmField
            val CODEC: StreamCodec<FriendlyByteBuf, RequestMapSnapshotC2S> = StreamCodec.of(
                { buf, v ->
                    buf.writeVarInt(v.minX)
                    buf.writeVarInt(v.minZ)
                    buf.writeVarInt(v.maxX)
                    buf.writeVarInt(v.maxZ)
                },
                { buf -> RequestMapSnapshotC2S(buf.readVarInt(), buf.readVarInt(), buf.readVarInt(), buf.readVarInt()) }
            )

            @JvmStatic
            fun handle(payload: RequestMapSnapshotC2S, context: IPayloadContext) {
                val sp = context.player() as ServerPlayer
                val cx = kotlin.math.round(sp.x).toInt()
                val cz = kotlin.math.round(sp.z).toInt()

                val radiusChunks: Int = try {
                    val cfg = net.shiroha233.roadweaver.config.ConfigService.get()
                    if (cfg.dynamicPlanEnabled()) cfg.dynamicPlanRadiusChunks() else cfg.initialPlanRadiusChunks()
                } catch (_: Throwable) {
                    256
                }

                val radiusBlocks = kotlin.math.max(1, radiusChunks) * 16

                CompletableFuture
                    .supplyAsync(
                        {
                            MapDataCollector.build(
                                sp.serverLevel(),
                                payload.minX,
                                payload.minZ,
                                payload.maxX,
                                payload.maxZ,
                                cx,
                                cz,
                                radiusBlocks
                            )
                        },
                        ComputeService.executor()
                    )
                    .thenAccept { snapshot ->
                        context.enqueueWork { PacketDistributor.sendToPlayer(sp, MapSnapshotS2C(snapshot)) }
                    }
            }
        }
    }

    data class MapSnapshotS2C(val snapshot: MapSnapshot) : CustomPacketPayload {
        override fun type(): CustomPacketPayload.Type<out CustomPacketPayload> = TYPE

        companion object {
            @JvmField
            val TYPE: CustomPacketPayload.Type<MapSnapshotS2C> = CustomPacketPayload.Type(SNAP_ID)

            @JvmField
            val CODEC: StreamCodec<FriendlyByteBuf, MapSnapshotS2C> = StreamCodec.of(
                { buf, v -> MapSnapshotCodec.write(buf, v.snapshot) },
                { buf -> MapSnapshotS2C(MapSnapshotCodec.read(buf)) }
            )

            @JvmStatic
            fun handle(payload: MapSnapshotS2C, context: IPayloadContext) {
                context.enqueueWork {
                    val mc = Minecraft.getInstance()
                    val screen = mc.screen
                    if (screen is RoadMapScreen) {
                        screen.setSnapshot(payload.snapshot)
                    }
                }
            }
        }
    }

    data class TeleportC2S(val x: Int, val y: Int, val z: Int) : CustomPacketPayload {
        override fun type(): CustomPacketPayload.Type<out CustomPacketPayload> = TYPE

        companion object {
            @JvmField
            val TYPE: CustomPacketPayload.Type<TeleportC2S> = CustomPacketPayload.Type(TP_REQ_ID)

            @JvmField
            val CODEC: StreamCodec<FriendlyByteBuf, TeleportC2S> = StreamCodec.of(
                { buf, v ->
                    buf.writeVarInt(v.x)
                    buf.writeVarInt(v.y)
                    buf.writeVarInt(v.z)
                },
                { buf -> TeleportC2S(buf.readVarInt(), buf.readVarInt(), buf.readVarInt()) }
            )

            @JvmStatic
            fun handle(payload: TeleportC2S, context: IPayloadContext) {
                context.enqueueWork {
                    val sp = context.player() as ServerPlayer
                    val allowed = sp.isCreative || sp.hasPermissions(2)
                    if (!allowed) {
                        PacketDistributor.sendToPlayer(sp, TeleportAckS2C(false, 0, 0, 0))
                        return@enqueueWork
                    }

                    val level = sp.serverLevel()
                    level.getChunk(payload.x shr 4, payload.z shr 4)
                    var ty = level.getHeight(Heightmap.Types.MOTION_BLOCKING_NO_LEAVES, payload.x, payload.z)
                    ty = if (ty <= level.minBuildHeight) level.seaLevel + 1 else ty + 1

                    sp.teleportTo(level, payload.x + 0.5, ty.toDouble(), payload.z + 0.5, sp.yRot, sp.xRot)
                    PacketDistributor.sendToPlayer(sp, TeleportAckS2C(true, payload.x, ty, payload.z))
                }
            }
        }
    }

    data class TeleportAckS2C(val ok: Boolean, val x: Int, val y: Int, val z: Int) : CustomPacketPayload {
        override fun type(): CustomPacketPayload.Type<out CustomPacketPayload> = TYPE

        companion object {
            @JvmField
            val TYPE: CustomPacketPayload.Type<TeleportAckS2C> = CustomPacketPayload.Type(TP_ACK_ID)

            @JvmField
            val CODEC: StreamCodec<FriendlyByteBuf, TeleportAckS2C> = StreamCodec.of(
                { buf, v ->
                    buf.writeBoolean(v.ok)
                    buf.writeVarInt(v.x)
                    buf.writeVarInt(v.y)
                    buf.writeVarInt(v.z)
                },
                { buf -> TeleportAckS2C(buf.readBoolean(), buf.readVarInt(), buf.readVarInt(), buf.readVarInt()) }
            )

            @JvmStatic
            fun handle(payload: TeleportAckS2C, context: IPayloadContext) {
                context.enqueueWork {
                    val mc = Minecraft.getInstance()
                    val player = mc.player ?: return@enqueueWork
                    if (payload.ok) {
                        player.displayClientMessage(
                            Component.translatable(
                                "gui.roadweaver.map.teleport.success_pos",
                                payload.x,
                                payload.y,
                                payload.z
                            ),
                            true
                        )
                    } else {
                        player.displayClientMessage(Component.translatable("gui.roadweaver.map.teleport.denied"), true)
                    }
                }
            }
        }
    }

    data class ManualConnectC2S(val ax: Int, val az: Int, val bx: Int, val bz: Int) : CustomPacketPayload {
        override fun type(): CustomPacketPayload.Type<out CustomPacketPayload> = TYPE

        companion object {
            @JvmField
            val TYPE: CustomPacketPayload.Type<ManualConnectC2S> = CustomPacketPayload.Type(MAN_REQ_ID)

            @JvmField
            val CODEC: StreamCodec<FriendlyByteBuf, ManualConnectC2S> = StreamCodec.of(
                { buf, v ->
                    buf.writeVarInt(v.ax)
                    buf.writeVarInt(v.az)
                    buf.writeVarInt(v.bx)
                    buf.writeVarInt(v.bz)
                },
                { buf -> ManualConnectC2S(buf.readVarInt(), buf.readVarInt(), buf.readVarInt(), buf.readVarInt()) }
            )

            @JvmStatic
            fun handle(payload: ManualConnectC2S, context: IPayloadContext) {
                context.enqueueWork {
                    val sp = context.player() as ServerPlayer
                    val level = sp.serverLevel()
                    val provider = WorldDataProvider.getInstance()
                    val origin = provider.getStructureConnections(level)
                    val list = if (origin != null) ArrayList(origin) else ArrayList()

                    val a = BlockPos(payload.ax, 0, payload.az)
                    val b = BlockPos(payload.bx, 0, payload.bz)

                    var exists = false
                    for (c in list) {
                        val f = c.from
                        val t = c.to
                        if ((f == a && t == b) || (f == b && t == a)) {
                            exists = true
                            break
                        }
                    }

                    if (!exists) {
                        list.add(Records.StructureConnection(a, b, Records.ConnectionStatus.PLANNED))
                        provider.setStructureConnections(level, list)
                    }
                }
            }
        }
    }
}
